tests/python/pants_test/engine/test_scheduler.py

```diff
@@ -61,6 +84,16 @@ def test_use_params(self):
     with self.assertRaises(Exception):
       self.scheduler.product_request(str, [Params(a)])
 
+  def test_transitive_params(self):
+    a, c = A(), C()
+    result_str, = self.scheduler.product_request(str, [Params(a, c)])
```

Comment:
How does this work? I didn't think we had rules returning `str` at all anywhere, because we use newtypes for everything?

Reply:
This is purely from the `consumes_a_and_b()` rule defined in this file -- having stringly typed products would be a truly horrid thing (imho).